src/lib.rs

```diff
@@ -0,0 +1,631 @@
+    uprotocol::{
+        Data, UAttributes, UCode, UEntity, UMessage, UMessageType, UPayload, UPayloadFormat,
+        UPriority, UStatus, UUri, Uuid,
+    },
+    uri::{
+        serializer::{MicroUriSerializer, UriSerializer},
+        validator::UriValidator,
+    },
+};
+use zenoh::{
+    config::Config,
+    prelude::{r#async::*, Sample},
+    queryable::{Query, Queryable},
+    sample::AttachmentBuilder,
+    subscriber::Subscriber,
+};
+
+pub struct ZenohListener {}
+pub struct ULinkZenoh {
+    session: Arc<Session>,
+    subscriber_map: Arc<Mutex<HashMap<String, Subscriber<'static, ()>>>>,
+    queryable_map: Arc<Mutex<HashMap<String, Queryable<'static, ()>>>>,
+    query_map: Arc<Mutex<HashMap<String, Query>>>,
+    callback_counter: AtomicU64,
+}
+
+impl ULinkZenoh {
+    /// # Errors
+    /// Will return `Err` if unable to create Zenoh session
+    pub async fn new(config: Config) -> Result<ULinkZenoh, UStatus> {
+        let Ok(session) = zenoh::open(config).res().await else {
+            return Err(UStatus::fail_with_code(
+                UCode::Internal,
+                "Unable to open Zenoh session",
+            ));
+        };
+        Ok(ULinkZenoh {
+            session: Arc::new(session),
+            subscriber_map: Arc::new(Mutex::new(HashMap::new())),
+            queryable_map: Arc::new(Mutex::new(HashMap::new())),
+            query_map: Arc::new(Mutex::new(HashMap::new())),
+            callback_counter: AtomicU64::new(0),
+        })
+    }
+
+    fn to_zenoh_key_string(uri: &UUri) -> Result<String, UStatus> {
+        let micro_uuri = MicroUriSerializer::serialize(uri).map_err(|_| {
+            UStatus::fail_with_code(
+                UCode::InvalidArgument,
+                "Unable to serialize into micro format",
+            )
+        })?;
+        let micro_zenoh_key: String = micro_uuri
+            .iter()
+            .fold(String::new(), |s, c| s + &format!("{c:02x}"));
+        Ok(micro_zenoh_key)
+    }
+
+    #[allow(clippy::match_same_arms)]
+    fn map_zenoh_priority(upriority: UPriority) -> Priority {
+        match upriority {
+            UPriority::UpriorityCs0 => Priority::Background,
+            UPriority::UpriorityCs1 => Priority::DataLow,
+            UPriority::UpriorityCs2 => Priority::Data,
+            UPriority::UpriorityCs3 => Priority::DataHigh,
+            UPriority::UpriorityCs4 => Priority::InteractiveLow,
+            UPriority::UpriorityCs5 => Priority::InteractiveHigh,
+            UPriority::UpriorityCs6 => Priority::RealTime,
+            // If uProtocol prioritiy isn't specified, use CS1(DataLow) by default.
+            // https://github.com/eclipse-uprotocol/uprotocol-spec/blob/main/basics/qos.adoc
+            UPriority::UpriorityUnspecified => Priority::DataLow,
+        }
+    }
+
+    // TODO: We need a standard way in uprotocol-rust to change UUID to String
+    fn uuid_to_string(uuid: &Uuid) -> String {
```

Comment:
There's a method to do this [over here](https://github.com/eclipse-uprotocol/up-rust/blob/main/src/proto/uprotocol/uuid.rs#L43) defined on `Uuid` that can be used. Might've been missed because we rename `Uuid` => `uproto_Uuid` [internal to this file](https://github.com/eclipse-uprotocol/up-rust/blob/f9163f7bc742d6d59b05b35a03666aaaab3cf80d/src/proto/uprotocol/uuid.rs#L17C31-L17C42). :sweat_smile:

Reply:
Yes, I missed that... Thanks!